File: examples/rs/benchmark_utils.rs
// Benchmark utilities for examples
// Provides timing and memory tracking capabilities

use std::time::Instant;

#[derive(Debug, Clone)]
pub struct BenchmarkResult {
    pub duration_ms: f64,
    pub memory_info: MemoryInfo,
}

#[derive(Debug, Clone)]
pub struct MemoryInfo {
    pub description: String,
}

impl MemoryInfo {
    pub fn best_effort() -> Self {
        // Best-effort memory tracking
        // On Linux, we could read /proc/self/status for RSS
        // For cross-platform simplicity, we provide a placeholder
        #[cfg(target_os = "linux")]
        {
            if let Ok(status) = std::fs::read_to_string("/proc/self/status") {
                for line in status.lines() {
                    if line.starts_with("VmRSS:") {
                        let parts: Vec<&str> = line.split_whitespace().collect();
                        if parts.len() >= 2 {
                            return MemoryInfo {
                                description: format!("RSS: {} kB", parts[1]),
                            };
                        }
                    }
                }
            }
        }
        
        MemoryInfo {
            description: "Memory tracking unavailable".to_string(),
        }
    }
}

pub struct Benchmark {
    start: Instant,
    label: String,
}

impl Benchmark {
    pub fn start(label: &str) -> Self {
        Self {
            start: Instant::now(),
            label: label.to_string(),
        }
    }
    
    pub fn finish(self) -> BenchmarkResult {
        let duration = self.start.elapsed();
        let duration_ms = duration.as_secs_f64() * 1000.0;
        let memory_info = MemoryInfo::best_effort();
        
        BenchmarkResult {
            duration_ms,
            memory_info,
        }
    }
    
    pub fn finish_and_print(self) -> BenchmarkResult {
        let result = self.finish();
        println!("⏱️  Runtime: {:.3}ms", result.duration_ms);
        println!("💾 Memory: {}", result.memory_info.description);
        result
    }
}

/// Format a section header
pub fn print_header(title: &str) {
    println!("\n{}", "═".repeat(60));
    println!("  {}", title);
    println!("{}\n", "═".repeat(60));
}

/// Format a subsection
pub fn print_section(title: &str) {
    println!("\n{}", "─".repeat(60));
    println!("{}", title);
    println!("{}\n", "─".repeat(60));
}

File: examples/py/benchmark_utils.py
"""Benchmark utilities for Python examples.

Provides timing and memory tracking capabilities using time.perf_counter()
and tracemalloc for memory allocation statistics.
"""

import time
import tracemalloc
from typing import Optional, Dict, Any


class BenchmarkResult:
    """Result of a benchmark measurement."""
    
    def __init__(self, duration_ms: float, memory_info: Dict[str, Any]):
        self.duration_ms = duration_ms
        self.memory_info = memory_info
    
    def __repr__(self):
        return f"BenchmarkResult(duration_ms={self.duration_ms:.3f}, memory={self.memory_info})"


class Benchmark:
    """Context manager for benchmarking code execution."""
    
    def __init__(self, label: str = "Benchmark"):
        self.label = label
        self.start_time: Optional[float] = None
        self.start_memory: Optional[tuple] = None
        self.result: Optional[BenchmarkResult] = None
    
    def __enter__(self):
        """Start timing and memory tracking."""
        tracemalloc.start()
        self.start_time = time.perf_counter()
        self.start_memory = tracemalloc.get_traced_memory()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Stop timing and memory tracking."""
        end_time = time.perf_counter()
        current_memory, peak_memory = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        
        duration_ms = (end_time - self.start_time) * 1000
        
        # Calculate memory delta
        start_current, start_peak = self.start_memory
        current_delta = current_memory - start_current
        peak_delta = peak_memory - start_peak
        
        self.result = BenchmarkResult(
            duration_ms=duration_ms,
            memory_info={
                'current_kb': current_delta / 1024,
                'peak_kb': peak_delta / 1024,
            }
        )
        return False
    
    def print_result(self):
        """Print the benchmark result."""
        if self.result:
            print(f"⏱️  Runtime: {self.result.duration_ms:.3f}ms")
            print(f"💾 Memory: Current: {self.result.memory_info['current_kb']:.2f} KB, "
                  f"Peak: {self.result.memory_info['peak_kb']:.2f} KB")


def print_header(title: str):
    """Print a formatted header."""
    print(f"\n{'═' * 60}")
    print(f"  {title}")
    print(f"{'═' * 60}\n")


def print_section(title: str):
    """Print a formatted section."""
    print(f"\n{'─' * 60}")
    print(title)
    print(f"{'─' * 60}\n")


def print_dist_table(rows, show_type: bool = False):
    """Print a compact, terminal-width-safe table of distribution stats.

    Parameters
    ----------
    rows : list of tuples
        Each tuple is one of:
          (name, dist)
          (name, dist, note)         — note printed on sub-line below the row
          (name, dist, note, type_str) — note + explicit type string column
        When show_type=False the type_str (if present) is ignored.
    show_type : bool
        When True, insert a 'Type' column between Var and the numeric stats.
        Pass the type string as ``row[3]`` (or it defaults to empty).
    """
    COL_VAR  = 8
    COL_TYPE = 12

    if show_type:
        header = (
            f"  {'Var':<{COL_VAR}}  {'Type':<{COL_TYPE}}"
            f"  {'Mean':>9}  {'Std':>9}  {'p5':>9}  {'p50':>9}  {'p95':>9}"
        )
        note_indent = " " * (2 + COL_VAR + 2 + COL_TYPE + 2 + 4)
    else:
        header = (
            f"  {'Var':<{COL_VAR}}"
            f"  {'Mean':>9}  {'Std':>9}  {'p5':>9}  {'p50':>9}  {'p95':>9}"
        )
        note_indent = " " * (2 + COL_VAR + 2 + 4)

    sep = "  " + "─" * (len(header) - 2)
    print(header)
    print(sep)

    for row in rows:
        name     = row[0]
        dist     = row[1]
        note     = row[2] if len(row) > 2 else ""
        type_str = row[3] if len(row) > 3 else ""

        stats = (
            f"  {dist.mean:9.4f}  {dist.std:9.4f}"
            f"  {dist.p5:9.4f}  {dist.p50:9.4f}  {dist.p95:9.4f}"
        )

        if show_type:
            print(f"  {name:<{COL_VAR}}  {type_str:<{COL_TYPE}}{stats}")
        else:
            print(f"  {name:<{COL_VAR}}{stats}")

        if note:
            print(f"{note_indent}└ {note}")
